src/pseudo/PspUpf.jl

```diff
@@ -51,150 +74,131 @@ function parse_upf_file(path; identifier=path)
             pswfc["angular_momentum"] == l
         end
         map(pswfc -> pswfc["radial_function"], pswfcs_l)
     end
+    
     pswfc_occs = map(0:lmax - 1) do l
         pswfcs_l = filter(pseudo["atomic_wave_functions"]) do pswfc
             pswfc["angular_momentum"] == l
         end
         map(pswfc -> pswfc["occupation"], pswfcs_l)
     end
-
-    # Currently not used; can be used for initializing the starting guess density.
+    
     rhoatom = pseudo["total_charge_density"]
 
-    # Useful precomputed quantities (q-independent parts of Fourier transform integrands)
-    rvlocpzdr = (rgrid .* vloc .+ Zion) .* drgrid
-    r2projsdr = map(projs) do projs_l
-        map(projs_l) do proj
+    ###
+    ### Calculated quantities
+    ###
+    vloc_interp = linear_interpolation((rgrid, ), vloc)
+    
+    projs_interp = map(projs) do projs_l
+        @views map(projs_l) do proj
+            ir_cut = lastindex(proj)
+            linear_interpolation((rgrid[1:ir_cut], ), proj)
+        end
+    end
+
+    r_vloc_corr_dr = (rgrid .* vloc .+ Zion) .* drgrid
+    
+    r2_projs_dr = map(projs) do projs_l
+        @views map(projs_l) do proj
             ir_cut = lastindex(proj)
             rgrid[1:ir_cut] .* proj .* drgrid[1:ir_cut]
         end
     end
 
-    PspUpf(Zion, lmax, rgrid, drgrid, vloc, vloc_interp, projs, projs_interp, h, pswfcs,
-           pswfc_occs, rhoatom, rvlocpzdr, r2projsdr; identifier)
+    PspUpf(Zion, lmax, rgrid, drgrid, vloc, projs, h, pswfcs, pswfc_occs, rhoatom,
+           vloc_interp, projs_interp, r_vloc_corr_dr, r2_projs_dr; identifier)
 end
 
 charge_ionic(psp::PspUpf) = psp.Zion
 
 """
-    eval_psp_projector_real(psp::PspUpf, i::Number, l::Number, r::Number)
+    eval_psp_projector_real(psp::PspUpf, i::Number, l::Number, r<:Real)
 
 Evaluate the ith Kleinman-Bylander β projector with angular momentum l at real-space
 distance r via linear interpolation on the real-space mesh.
 
-Note: UPFs store `r[i] p_{il}(r[i])`, so r=0 is undefined and will error.
+Note: UPFs store `r[j] β_{li}(r[j])`, so r=0 is undefined and will error.
 """
 function eval_psp_projector_real(psp::PspUpf, i, l, r::T) where {T <: Real}
     psp.projs_interp[l+1][i](r) / r
 end
 
 """
-    eval_psp_projector_fourier(psp::PspUPF, i::Number, l::Number, q::Number)
-
-
-Evaluate the radial part of the i-th projector for angular momentum l at the
-reciprocal vector with modulus q:
-p(q) = ∫R^3 p{il}(r) e^{-iqr} dr
-     = 4π ∫{R+} r^2 p_{il}(r) j_l(q r) dr
-     = 4π Σ{i} r[i]^2 p_{il}(r[i]) j_l(q r[i]) dr[i]
+    eval_psp_projector_fourier(psp::PspUPF, i::Number, l::Number, q<:Real)
 
-Note: UPFs store `r[i] p_{il}(r[i])`
+For UPFs, the integral is transformed to the following sum:
+4π Σ{k} j_l(q r[k]) (r[k]^2 p_{il}(r[k]) dr[k])
 """
-function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T <: Real}
+function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T) where {T <: Real}
```

Comment:
You checked this infers the type correctly? We had perf issues with that in the HGHs

Reply:
Ah no I didn't, it did pop into my mind when I was deleting it though. I should put it back, it does cause performance issues.